features/steps/thens/existance.py

```diff
@@ -22,3 +24,14 @@ def step_impl(context, inst, constraint, num, entity):
 
         if not op(len(instances_in_model), num):
             yield StepResult(expected=num, observed=len(instances_in_model))
+
+
+@validate_step('There must be {num:d} representation item(s)')
+def step_impl(context, inst, num):
+    # inst is a list that always has one item but that item may contain multiple tuples
```

Comment:
I don't think we can get away with it like that. The structure comes from the IFC data model. If we print the instances after every given statement and interleave these line by line we get the following structure:

```
#2288=IfcAlignmentSegment('3FHEGq6711TulOKVgmhD$$',#1,'h1',$,$,#2289,#2458,#2293),
    #2458=IfcProductDefinitionShape($,$,(#2460)),
        (#2460=IfcShapeRepresentation(#2459,'Axis','Segment',(#2450)),),
                ((#2450),),
#2295=IfcAlignmentSegment('2NHsvN8xnFnetsLAyLgoE6',#1,'h2',$,$,#2296,#2469,#2300),
    #2469=IfcProductDefinitionShape($,$,(#2470)),
        (#2470=IfcShapeRepresentation(#2459,'Axis','Segment',(#2461,#24610)),),
                ((#2461,#24610),),
#2302=IfcAlignmentSegment('1$forlm9r0XeFCaEQu3k18',#1,'h3',$,$,#2303,#2479,#2307),
    #2479=IfcProductDefinitionShape($,$,(#2480)),
        (#2480=IfcShapeRepresentation(#2459,'Axis','Segment',(#2471)),),
                ((#2471,),)
#2309=IfcAlignmentSegment('3gVJrcRzD8yPNgP0ikTfG9',#1,'h4',$,$,#2310,#2489,#2314),
    #2489=IfcProductDefinitionShape($,$,(#2490)),
        (#2490=IfcShapeRepresentation(#2459,'Axis','Segment',(#2481)),),
                ((#2481,),)
```

So `len(inst)` is actually the number of shape representations, `len(inst[0])` is the amount of items for the first shape representation.

--> So, quick and dirty way to move on would be to change it to `all(len(v) == num for v in inst)`. All shape representations should have length of `num`.

But maybe we should also think about what this means for the semantics of this rule.

Are multiple representations allowed? Are representations with a different identifier allowed to have multiple segments?

Is the rule:

a)

- All representation identifiers should be Segment
- All sequences of items should be length=1

(this allows for the case of a segment without a representation)

b)

- An alignmentseg should have a at least one Segment representation
- For the representation that has Id=Segment only a single item is allowed

(this disallows a segment without a representation - I don't think we can do that given businesslogic-only exchanges)

or c)

- If an alignmentseg has a representation, then
- it should have a at least one Segment representation
- and then representation should have a sequence of items should be length=1

This is a good example of where a simple rule gets complex due to the variability of the IFC schema.

My recommendation:

- Stick with option (a) this is closest to what's currently implemented. I don't foresee other representation Ids on segments for now (but it is a bit against the nature of IFC to forbid supplementary reps, we can move to ) The case of a segment without a representation should be allowed.
- Create some more tests:
  - Segment without representation
  - Segment with two representations

Reply:
The implementation of scenario 2 has been adjusted.  I used a for loop so that we could extract the actual number of items and report that back as "observed".  I also added more tests for the two cases listed above.

Reply:
For a separate PR perhaps, but now it's fresh in my head. 
We could use this (as well as similar step_impl functions in `values.py`) more with reuse in mind, so for example 
```
      Given Its attribute Items
      Then  It must be of length [{pred}]{num:d}
```
  
  pred (abbreviation for 'predicament'?) can then be one of: 
https://github.com/buildingSMART/ifc-gherkin-rules/blob/f6d10cf03d0acc7a0a94383185402e642803e114/features/steps/utils/misc.py#L104-L117
Used in the step_impl function https://github.com/buildingSMART/ifc-gherkin-rules/blob/f6d10cf03d0acc7a0a94383185402e642803e114/features/steps/thens/existance.py#L20-L30
  
` [{pred}]` is written in brackets for behave to parse match correctly. 
 An alternative would be using a `register_type`, but I'm not sure how to do this in this case. 
 
 But, as said, it would probably be better to do this outside of this work.